### ISSUE TITLE
Umbrella: 99/380 numeric divergence cases under fullgraph=False (backend=eager, NGB=False)

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#128]
## TL;DR

Under `torch.compile(backend="eager", fullgraph=False)` with no nested-graph-breaks flag, **99 of 380 (26.1%) (model, mode) pair-rows show numeric divergence** vs pure eager. This is unexpected: `backend="eager"` is a no-op compile (FX graph capture → re-execute as eager) and should not introduce numeric drift.

This is an umbrella to track the investigation. Single per-model issues will be filed once root causes are isolated.

## Sweep that surfaced this

- **Run:** `experiments/results/ngb-baseline-2026-05-15-20260515-192617/`
- **Config:** `--compile-kwargs '{"fullgraph": false}'`, NO `--dynamo-config nested_graph_breaks=True` (NGB defaults to False)
- **Backend:** `eager` (default; passed implicitly via dynamo)
- **Stack:** torch 2.13.0.dev20260502+cu126, transformers 5.6.2, diffusers 0.38.0
- **Cohort:** 190 models × 2 modes (eval + train) = 380 pair-rows
- **Pass:** identify (numeric verify is populated by identify pass, not explain pass)

## Summary table

| mode  | total | match | divergence | divergence % |
|-------|-------|-------|------------|--------------|
| eval  | 190   | 146   | 44         | 23.2%        |
| train | 190   | 134   | 55         | 28.9%        |
| **all** | **380** | **280** | **99**   | **26.1%**    |

(train mode also has 1 None row — a create_error.)

### Divergence classification

| class                                | count | description                                                                 |
|--------------------------------------|-------|-----------------------------------------------------------------------------|
| Noise-floor dominant (`numeric_noise_floor_dominant=True`) | 63 | Less-flaky retry matches under norm-eps amplification — likely harness / non-determinism. |
| Structural (`numeric_noise_floor_dominant=False`) | 36 | Real numeric differences exceeding the noise floor.                          |

Of the 36 structural divergences:
- **14 large-structural** (max_diff 1.86 to 5.92, all train-mode) — speech / encoder-decoder models (`Wav2Vec2`, `Hubert`, `UniSpeech`, `UniSpeechSat`, `WavLM`, `Data2VecAudio`, `Wav2Vec2Conformer`, `SEW`, `Reformer`, `M2M100`, `M2M100ForConditionalGeneration`, `PLBart`, `PLBartForConditionalGeneration`, `SpeechEncoderDecoder`).
- **22 small-structural** (max_diff 8.58e-6 to 2.72e-5) — multimodal / MoE / hybrid: `Gemma4ForConditionalGeneration` (both modes), `FalconH1ForCausalLM` (eval), Voxtral / Glm4v / Glm4vMoe / Glm46V / Qwen2VL / Qwen2_5_VL / Qwen2_5OmniThinker / Qwen3_5 / Ernie4_5_VL_Moe / Ernie4_5_VLMoe families.

## Why this is surprising

`torch.compile(backend="eager")` is intended as a no-op compile: dynamo traces an FX graph, then the "eager" backend re-executes that graph using stock PyTorch ops. There is no Inductor codegen, no kernel fusion, no decomposition into different reduction order. In principle the output should be bitwise-identical to pure eager (modulo any side-effects of tracing/guards/functionalization).

Hypotheses to investigate (each tested in the model-by-model walk):

1. **Guard / dispatcher side-effects** reorder summation in attention / reduction ops.
2. **Pre-grad decompositions** even when backend is eager (e.g., AOTAutograd functionalization).
3. **In-place → out-of-place rewrite** changes memory aliasing and reduction order.
4. **RNG state advancement** differs at graph-break re-entry boundaries.
5. **Autocast scope** boundaries shift across the compile boundary.
6. **Tensor contiguity / stride** changes through the graph.

## Investigation plan

Tracked in `PLAN.md` (WS2 Task 8). High level:

1. **Picked first deep-dive candidate:** `Gemma4ForConditionalGeneration` (eval mode, structural, max_diff = 2.605e-05 — largest of the 12 eval-mode structural cases; smallest GPU footprint at 7.0 GB; eager time 2.43 s).
2. **Apples-to-apples minimal repro:** strip the sweep harness; run model under (a) pure eager and (b) `torch.compile(backend="eager", fullgraph=False)` with identical seed / inputs / dtype. Validate divergence is real (not harness artifact).
3. **Op / layer isolation:** register forward-hook diff probes at `nn.Module` boundaries; compare compact stats (shape, dtype, mean, max_abs, first sample, nan_count) eager vs compile(eager); bisect to first divergent boundary; drill into sub-stages.
4. **Cross-model class triage:** once one root cause is known, check whether the same mechanism explains other structural divergences in the same architectural family.

## Methodology references

The model-by-model isolation walk follows a standard intermediate-stats bisection recipe:

- Same prompt / input ids / seed / dtype on both paths.
- Compact stats first (shape, dtype, device, mean, max_abs, first sample, nan_count) — full tensors only when stage-level stats identify the first divergent boundary.
- Register hooks via `module.register_forward_hook(...)` and remove handles in `try`/`finally`.
- Stage filters: `layer_input`, attention input/output, post-attention residual, pre-FFN norm, raw FFN output, post-FFN residual.
- For MoE, drill into routing, expert GEMM, post-expert-norm, reduce-back.

## Data files

- Results: `experiments/results/ngb-baseline-2026-05-15-20260515-192617/results.jsonl`
- Streaming: `experiments/results/ngb-baseline-2026-05-15-20260515-192617/identify_streaming.jsonl`
- Config: `experiments/results/ngb-baseline-2026-05-15-20260515-192617/config.json`
- Sweep state: `experiments/results/ngb-baseline-2026-05-15-20260515-192617/sweep_state.json`

## Full list of 36 structural divergences

Sorted by max_diff descending:

| # | model | mode | max_diff | first_divergence |
|---|-------|------|----------|------------------|
| 1 | ReformerModel | train | 5.92e+00 | last_hidden_state |
| 2 | UniSpeechSatModel | train | 5.64e+00 | last_hidden_state |
| 3 | UniSpeechModel | train | 5.33e+00 | last_hidden_state |
| 4 | Wav2Vec2Model | train | 5.23e+00 | last_hidden_state |
| 5 | SpeechEncoderDecoderModel | train | 5.04e+00 | logits |
| 6 | Data2VecAudioModel | train | 5.00e+00 | last_hidden_state |
| 7 | HubertModel | train | 4.96e+00 | last_hidden_state |
| 8 | WavLMModel | train | 4.79e+00 | last_hidden_state |
| 9 | M2M100Model | train | 4.10e+00 | encoder_last_hidden_state |
| 10 | PLBartModel | train | 3.82e+00 | last_hidden_state |
| 11 | Wav2Vec2ConformerModel | train | 3.35e+00 | last_hidden_state |
| 12 | M2M100ForConditionalGeneration | train | 2.43e+00 | logits |
| 13 | PLBartForConditionalGeneration | train | 2.00e+00 | encoder_last_hidden_state |
| 14 | SEWModel | train | 1.86e+00 | last_hidden_state |
| 15 | Gemma4ForConditionalGeneration | train | 2.72e-05 | logits |
| 16 | Gemma4ForConditionalGeneration | eval  | 2.61e-05 | logits |
| 17 | FalconH1ForCausalLM | eval  | 2.03e-05 | logits |
| 18 | VoxtralRealtimeForConditionalGeneration | eval  | 1.76e-05 | logits |
| 19 | VoxtralRealtimeForConditionalGeneration | train | 1.67e-05 | logits |
| 20 | Glm4vMoeForConditionalGeneration | train | 1.65e-05 | logits |
| 21 | Glm46VForConditionalGeneration | train | 1.57e-05 | logits |
| 22 | Glm4vForConditionalGeneration | eval  | 1.57e-05 | logits |
| 23 | Glm4vForConditionalGeneration | train | 1.53e-05 | logits |
| 24 | Qwen2_5OmniThinkerForConditionalGeneration | eval  | 1.53e-05 | logits |
| 25 | Qwen2_5OmniThinkerForConditionalGeneration | train | 1.53e-05 | logits |
| 26 | Glm46VForConditionalGeneration | eval  | 1.50e-05 | logits |
| 27 | Glm4vMoeForConditionalGeneration | eval  | 1.48e-05 | logits |
| 28 | Qwen2_5_VLForConditionalGeneration | eval  | 1.41e-05 | logits |
| 29 | Qwen2VLForConditionalGeneration | eval  | 1.29e-05 | logits |
| 30 | Qwen2VLForConditionalGeneration | train | 1.29e-05 | logits |
| 31 | Ernie4_5_VLMoeForConditionalGeneration | train | 1.00e-05 | logits |
| 32 | Qwen3_5ForConditionalGeneration | train | 9.54e-06 | logits |
| 33 | Ernie4_5_VL_MoeForConditionalGeneration | train | 9.18e-06 | logits |
| 34 | Ernie4_5_VLMoeForConditionalGeneration | eval  | 9.06e-06 | logits |
| 35 | Ernie4_5_VL_MoeForConditionalGeneration | eval  | 9.06e-06 | logits |
| 36 | Qwen3_5ForConditionalGeneration | eval  | 8.58e-06 | logits |

Note: rows 15-36 all diverge at the final logits with max_diff in the 8.58e-6 to 2.72e-5 range — barely above the noise floor — and are all multimodal / MoE / hybrid models. Rows 1-14 are large-magnitude divergences (> 1.0) exclusively in train mode on speech / encoder-decoder models. The two clusters likely have different root causes.

**Comment by penguinwu** [comment]
## Root cause identified for the 22 small-structural multimodal/MoE cluster

The 22 small-structural divergences in this cluster (rows 15-36 in the original table, max_diff 8.58e-6 to 2.72e-5) share a single root cause: **HF transformers materializes an explicit boolean attention mask under dynamo tracing, which changes SDPA backend kernel selection.** Below is the walk.

### Step 1 — Apples-to-apples repro on Gemma4ForConditionalGeneration

Pure-Python harness-stripped repro (`/tmp/repro_gemma4.py`) with identical seed, inputs, dtype, model instance:

- `bitwise_equal=False`, `allclose(atol=1e-6,rtol=1e-4)=False`
- `max_diff=2.956e-05` (matches sweep's 2.605e-05 within nightly-drift)
- `image_hidden_states` bitwise EQUAL → divergence isolated to logits / text path

Confirms divergence is real, not a harness artifact.

### Step 2 — Op/layer isolation via nn.Module forward hooks

Hook script `/tmp/repro_gemma4_hooks.py` registers hooks on every named module of ONE shared model instance, runs eager then `torch.compile(backend='eager')`, diffs compact stats (shape/dtype/mean/abs_max/sum):

- 71 of 112 modules **IDENTICAL**, 41 divergent
- `model.vision_tower.*` entirely IDENTICAL (vision path is clean)
- `model.language_model.embed_tokens` IDENTICAL
- `model.language_model.rotary_emb` IDENTICAL
- Every `model.language_model.layers.*.self_attn.{q,k,v}_proj` output IDENTICAL
- `model.language_model.layers.*.self_attn` outputs DIVERGENT (Δabsmax 1e-7 to 1e-5)
- All downstream (o_proj, mlp, norms) inherit drift from the attention output

Conclusion: q/k/v projections produce identical tensors, but the attention computation (scaled-dot-product) between them diverges.

### Step 3 — SDPA backend hypothesis test

`/tmp/repro_gemma4_sdpa.py` forces a specific SDPA backend on BOTH eager and compile runs via `torch.nn.attention.sdpa_kernel(...)`:

| Backend | bitwise | max_diff |
|---|---|---|
| default | False | 3.42e-05 |
| **MATH** | **True** | **0.000e+00** |
| FLASH | n/a (kernel unavailable for this input shape) | |
| EFFICIENT | n/a | |
| CUDNN | n/a | |

Forcing MATH on both runs eliminates the divergence completely → confirms SDPA backend selection is the source.

### Step 4 — Why does the backend choice differ?

SDPA call-args trace (`/tmp/repro_gemma4_sdpa_trace.py`) monkey-patches `F.scaled_dot_product_attention` to log every call:

| arg | eager run | compile run |
|---|---|---|
| `is_causal` | True | **False** |
| `attn_mask` | None | **`(B,1,L,L)` torch.bool** |
| `enable_gqa` | True | **False** |
| K/V shape | `(2, 4, L, D)` | **`(2, 8, L, D)` (manually repeated)** |

Every SDPA call has different arguments. Backend=eager is **not** a no-op compile here — something in the model code is taking a different path under tracing.

### Step 5 — Locate the divergent branch in HF source

Mask-construction trace (`/tmp/repro_gemma4_mask_trace.py`) patches the `sdpa_mask` function in `transformers.masking_utils`. Even though `allow_is_causal_skip=True` is identical in both runs:

```
[eager]  call 1: q_length=17 kv_length=17 allow_is_causal_skip=True in_compile=False result=None
[compile] call 1: q_length=17 kv_length=17 allow_is_causal_skip=True in_compile=True  result=Tensor(2,1,17,17)_torch.bool
```

The materialized-mask branch is `_ignore_causal_mask_sdpa()`:

```python
if (
    not is_tracing(padding_mask)                                  # ← FALSE under dynamo tracing
    and (query_length == 1 or kv_length == query_length)
    and (local_attention_size is None or kv_length < local_attention_size)
    and (padding_mask is None or padding_mask.all())
):
    return True
return False
```

`is_tracing()` returns True under dynamo, so the `None` mask shortcut is skipped and a real `(B,1,L,L)` boolean mask gets built. The leading comment in the HF source explains:

> *"When using `torch.export` or `torch.onnx.dynamo_export`, we must pass an example input, and `is_causal` behavior is hard-coded to the forward... Thus, we only set `ignore_causal_mask = True` if we are not tracing."*

The branch is intentional, motivated by `torch.export` / ONNX correctness. `torch.compile(backend="eager")` triggers the same `is_tracing()` check and gets routed onto the same materialized-mask path.

### Step 6 — Downstream effect

Once an explicit mask is present:

1. `sdpa_attention_forward` calls `use_gqa_in_sdpa(attention_mask, key)`. That helper returns `torch>=2.5 and attention_mask is None`. With a mask present → False.
2. False routes through `repeat_kv(key, num_key_value_groups)` — manual K/V replication (4 KV heads → 8 to match Q heads), `enable_gqa=False`.
3. `is_causal` is recomputed as `query.shape[2] > 1 and attention_mask is None and is_causal` → False (because mask is now present).
4. `F.scaled_dot_product_attention` is then called with completely different args than in eager. The internal backend picker selects a different kernel for the new arg pattern → different reduction order / fused-vs-unfused arithmetic → ~1e-5 numerical drift.

### Step 7 — Cross-model triage on the cluster

`/tmp/cross_model_sdpa.py` runs the same `default vs force-MATH` comparison on 4 other models from the small-structural list:

| Model | default max_diff | MATH max_diff |
|---|---|---|
| FalconH1ForCausalLM | 2.07e-05 | 0.000e+00 (bitwise) |
| Qwen2VLForConditionalGeneration | 1.55e-05 | 0.000e+00 (bitwise) |
| Glm4vForConditionalGeneration | 1.43e-05 | 0.000e+00 (bitwise) |
| Qwen3_5ForConditionalGeneration | 1.00e-05 | 0.000e+00 (bitwise) |

Default values match each model's sweep-reported max_diff within nightly drift; forcing MATH on both paths yields bitwise equality for every one. **The 22 small-structural rows in this umbrella all share the same root cause.**

### Implications

- **`torch.compile(backend="eager")` is not actually a no-op compile** for any model that goes through HF's `sdpa_mask`/`sdpa_attention_forward` (i.e. most modern HF causal-LM and multimodal models). The arg-rewriting branch fires whenever `is_tracing()` is True.
- The numerical drift is small (~1e-5) because both eager and compile paths still go through the same SDPA op family, just different kernel branches that have small fused-vs-unfused reduction-order differences. None of the models in this cluster diverge in a structural / functional sense.
- The 14 LARGE-structural divergences in the original table (max_diff 1.86 to 5.92, all train-mode speech models) are a **different cluster** — the magnitudes are far too large to be explained by an SDPA kernel swap, and they only fire in train mode. Tracking separately as Task 8.6.

### Possible fixes (not prescribing, just enumerating)

- HF side: gate the `is_tracing()` check on `torch.jit.is_tracing()` only (i.e. export/onnx path), not on dynamo tracing. The HF comment carries a `TODO: cyril -> probably revisit and remove this`, suggesting the maintainers know this is fragile.
- PyTorch side: make SDPA's backend dispatcher prefer the same kernel for "is_causal=True + None mask + GQA=True" and "is_causal=False + bool mask + KV-repeated + GQA=False" when they are mathematically equivalent. This would make the dispatcher robust to HF's arg-rewrite.
- Workaround for users: wrap inference in `with torch.nn.attention.sdpa_kernel([SDPBackend.MATH]):` to force the bitwise-stable kernel on both paths.

### Repro artifacts

All scripts live in /tmp on the investigation VM:

- `repro_gemma4.py` — apples-to-apples logit diff
- `repro_gemma4_hooks.py` — nn.Module hook diff over all 112 modules
- `repro_gemma4_sdpa.py` — SDPA-backend forcing test
- `repro_gemma4_sdpa_trace.py` — SDPA call-args trace
- `repro_gemma4_mask_trace.py` — `sdpa_mask` trace pinpointing the `is_tracing()` branch
- `cross_model_sdpa.py` — cross-model triage on 4 representatives
